# Model Fit in Linear Regression - Lab

## Introduction
In this lab, you'll learn how to evaluate your model results and you'll learn how to select the appropriate features using stepwise selection.

## Objectives
You will be able to:
* Use stepwise selection methods to determine the most important features for a model
* Use recursive feature elimination to determine the most important features for a model

## The Boston Housing Data once more

We pre-processed the Boston Housing data the same way we did before:

- We dropped `'ZN'` and `'NOX'` completely
- We categorized `'RAD'` in 3 bins and `'TAX'` in 4 bins
- We transformed `'RAD'` and `'TAX'` to dummy variables and dropped the first variable
- We used min-max-scaling on `'B'`, `'CRIM'`, and `'DIS'` (and logtransformed all of them first, except `'B'`)
- We used standardization on `'AGE'`, `'INDUS'`, `'LSTAT'`, and `'PTRATIO'` (and logtransformed all of them first, except for `'AGE'`) 

In [1]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_boston
boston = load_boston()

boston_features = pd.DataFrame(boston.data, columns = boston.feature_names)
boston_features = boston_features.drop(['NOX', 'ZN'],axis=1)

# First, create bins for based on the values observed. 3 values will result in 2 bins
bins = [0,6,  24]
bins_rad = pd.cut(boston_features['RAD'], bins)
bins_rad = bins_rad.cat.as_unordered()

# First, create bins for based on the values observed. 4 values will result in 3 bins
bins = [0, 270, 360, 712]
bins_tax = pd.cut(boston_features['TAX'], bins)
bins_tax = bins_tax.cat.as_unordered()

tax_dummy = pd.get_dummies(bins_tax, prefix='TAX', drop_first=True)
rad_dummy = pd.get_dummies(bins_rad, prefix='RAD', drop_first=True)
boston_features = boston_features.drop(['RAD', 'TAX'], axis=1)
boston_features = pd.concat([boston_features, rad_dummy, tax_dummy], axis=1)

age = boston_features['AGE']
b = boston_features['B']
logcrim = np.log(boston_features['CRIM'])
logdis = np.log(boston_features['DIS'])
logindus = np.log(boston_features['INDUS'])
loglstat = np.log(boston_features['LSTAT'])
logptratio = np.log(boston_features['PTRATIO'])

# Min-Max scaling
boston_features['B'] = (b-min(b))/(max(b)-min(b))
boston_features['CRIM'] = (logcrim-min(logcrim))/(max(logcrim)-min(logcrim))
boston_features['DIS'] = (logdis-min(logdis))/(max(logdis)-min(logdis))

# Standardization
boston_features['AGE'] = (age-np.mean(age))/np.sqrt(np.var(age))
boston_features['INDUS'] = (logindus-np.mean(logindus))/np.sqrt(np.var(logindus))
boston_features['LSTAT'] = (loglstat-np.mean(loglstat))/np.sqrt(np.var(loglstat))
boston_features['PTRATIO'] = (logptratio-np.mean(logptratio))/(np.sqrt(np.var(logptratio)))

In [2]:
boston_features

CRIM     INDUS  CHAS     RM       AGE       DIS   PTRATIO         B  \
0    0.000000 -1.704344   0.0  6.575 -0.120013  0.542096 -1.443977  1.000000   
1    0.153211 -0.263239   0.0  6.421  0.367166  0.623954 -0.230278  1.000000   
2    0.153134 -0.263239   0.0  7.185 -0.265812  0.623954 -0.230278  0.989737   
3    0.171005 -1.778965   0.0  6.998 -0.809889  0.707895  0.165279  0.994276   
4    0.250315 -1.778965   0.0  7.147 -0.511180  0.707895  0.165279  1.000000   
..        ...       ...   ...    ...       ...       ...       ...       ...   
501  0.240099  0.410792   0.0  6.593  0.018673  0.331081  1.095518  0.987619   
502  0.206118  0.410792   0.0  6.120  0.288933  0.297277  1.095518  1.000000   
503  0.236926  0.410792   0.0  6.976  0.797449  0.274575  1.095518  1.000000   
504  0.298671  0.410792   0.0  6.794  0.736996  0.315551  1.095518  0.991301   
505  0.210954  0.410792   0.0  6.030  0.434732  0.335545  1.095518  1.000000   

        LSTAT  RAD_(6, 24]  TAX_(270, 360]  TAX_(360, 712]  
0   -1.275260            0               1               0  
1   -0.263711            0               0               0  
2   -1.627858            0               0               0  
3   -2.153192            0               0               0  
4   -1.162114            0               0               0  
..        ...          ...             ...             ...  
501 -0.169811            0               1               0  
502 -0.274682            0               1               0  
503 -1.067939            0               1               0  
504 -0.836660            0               1               0  
505 -0.510809            0               1               0  

[506 rows x 12 columns]

## Perform stepwise selection

The function for stepwise selection is copied below. Use this function provided on your preprocessed Boston Housing data.

In [14]:
import statsmodels.api as sm

def stepwise_selection(X, y, 
                       initial_list=[], 
                       threshold_in=0.01, 
                       threshold_out = 0.05, 
                       verbose=True):
    """ Perform a forward-backward feature selection 
    based on p-value from statsmodels.api.OLS
    Arguments:
        X - pandas.DataFrame with candidate features
        y - list-like with the target
        initial_list - list of features to start with (column names of X)
        threshold_in - include a feature if its p-value < threshold_in
        threshold_out - exclude a feature if its p-value > threshold_out
        verbose - whether to print the sequence of inclusions and exclusions
    Returns: list of selected features 
    Always set threshold_in < threshold_out to avoid infinite looping.
    See https://en.wikipedia.org/wiki/Stepwise_regression for the details
    """
    included = list(initial_list)
    while True:
        changed=False
        # forward step
        excluded = list(set(X.columns)-set(included))
        new_pval = pd.Series(index=excluded)
        for new_column in excluded:
            model = sm.OLS(y, sm.add_constant(pd.DataFrame(X[included+[new_column]]))).fit()
            new_pval[new_column] = model.pvalues[new_column]
        best_pval = new_pval.min()
        if best_pval < threshold_in:
            best_feature = new_pval.idxmin()
            included.append(best_feature)
            changed=True
            if verbose:
                print('Add  {:30} with p-value {:.6}'.format(best_feature, best_pval))

        # backward step
        model = sm.OLS(y, sm.add_constant(pd.DataFrame(X[included]))).fit()
        # use all coefs except intercept
        pvalues = model.pvalues.iloc[1:]
        worst_pval = pvalues.max() # null if pvalues is empty
        if worst_pval > threshold_out:
            changed=True
            worst_feature = pvalues.argmax()
            included.remove(worst_feature)
            if verbose:
                print('Drop {:30} with p-value {:.6}'.format(worst_feature, worst_pval))
        if not changed:
            break
    return included

In [15]:
# Your code here

# dependent variable
# boston.keys()
# boston.target
y = pd.DataFrame(boston.target, columns = ['price'])
y

results= stepwise_selection(boston_features, y, 
                       initial_list=[], 
                       threshold_in=0.01, 
                       threshold_out = 0.05, 
                       verbose=True)
#     """ Perform a forward-backward feature selection 
#     based on p-value from statsmodels.api.OLS
#     Arguments:
#         X - pandas.DataFrame with candidate features
#         y - list-like with the target
#         initial_list - list of features to start with (column names of X)
#         threshold_in - include a feature if its p-value < threshold_in
#         threshold_out - exclude a feature if its p-value > threshold_out
#         verbose - whether to print the sequence of inclusions and exclusions
#     Returns: list of selected features 
#     Always set threshold_in < threshold_out to avoid infinite looping.
#     See https://en.wikipedia.org/wiki/Stepwise_regression for the details
#     """

C:\Users\poari\Anaconda3\envs\learn-env\lib\site-packages\numpy\core\fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


Add  LSTAT                          with p-value 9.27989e-122
Add  RM                             with p-value 1.98621e-16
Add  PTRATIO                        with p-value 2.5977e-12
Add  DIS                            with p-value 2.85496e-09
Add  B                              with p-value 2.77572e-06
Add  INDUS                          with p-value 0.0017767
Add  CHAS                           with p-value 0.0004737


In [22]:
type(results)
results


['LSTAT', 'RM', 'PTRATIO', 'DIS', 'B', 'INDUS', 'CHAS']

In [88]:
# step by step
X=boston_features
initial_list=[], 
threshold_in=0.01, 
threshold_out = 0.05, 
verbose=True

included = []
# included

changed=False
# changed
#         # forward step

excluded = list(set(X.columns)-set(included))
print('excluded = ', excluded)

new_pval = pd.Series(index=excluded)



for new_column in excluded:
    model = sm.OLS(y, sm.add_constant(pd.DataFrame(X[included+[new_column]]))).fit()    
    new_pval[ new_column ] = model.pvalues[new_column] 
    print( 'new_pval[ ' , new_column , '] =' , model.pvalues[new_column] )

    best_pval = new_pval.min()
    if best_pval < threshold_in:
         best_feature = new_pval.idxmin()
#             included.append(best_feature)
#             changed=True
#             if verbose:
#                 print('Add  {:30} with p-value {:.6}'.format(best_feature, best_pval))

#         # backward step
#         model = sm.OLS(y, sm.add_constant(pd.DataFrame(X[included]))).fit()
#         # use all coefs except intercept
#         pvalues = model.pvalues.iloc[1:]
#         worst_pval = pvalues.max() # null if pvalues is empty
#         if worst_pval > threshold_out:
#             changed=True
#             worst_feature = pvalues.argmax()
#             included.remove(worst_feature)
#             if verbose:
#                 print('Drop {:30} with p-value {:.6}'.format(worst_feature, worst_pval))
#         if not changed:
#             break
#     return included

excluded =  ['INDUS', 'LSTAT', 'CHAS', 'TAX_(270, 360]', 'RM', 'DIS', 'RAD_(6, 24]', 'TAX_(360, 712]', 'CRIM', 'B', 'PTRATIO', 'AGE']
new_pval[  INDUS ] = 2.779090873336999e-36
new_pval[  LSTAT ] = 9.279885665959799e-122
new_pval[  CHAS ] = 7.390623170519864e-05
new_pval[  TAX_(270, 360] ] = 0.0397110257887882
new_pval[  RM ] = 2.48722887100781e-74
new_pval[  DIS ] = 1.996983855981312e-11
new_pval[  RAD_(6, 24] ] = 1.9463524894118995e-08
new_pval[  TAX_(360, 712] ] = 2.709643599750018e-19
new_pval[  CRIM ] = 3.898392697488064e-27
new_pval[  B ] = 1.3181127340754786e-14
new_pval[  PTRATIO ] = 7.910742574115878e-34
new_pval[  AGE ] = 1.5699822091880774e-18


In [87]:
X[['CHAS','DIS']]

CHAS       DIS
0     0.0  0.542096
1     0.0  0.623954
2     0.0  0.623954
3     0.0  0.707895
4     0.0  0.707895
..    ...       ...
501   0.0  0.331081
502   0.0  0.297277
503   0.0  0.274575
504   0.0  0.315551
505   0.0  0.335545

[506 rows x 2 columns]

### Build the final model again in Statsmodels

In [ ]:
# Your code here

The stepwise procedure mentions that `'INDUS'` was added with a p-value of 0.0017767, but our statsmodels output returns a p-value of 0.000. Use some of the stepwise procedure logic to find the intuition behind this!

## Use Feature ranking with recursive feature elimination

Use feature ranking to select the 5 most important features

In [ ]:
# Your code here

Fit the linear regression model again using the 5 selected columns

In [ ]:
# Your code here

Now, predict $\hat y$ using your model. You can use `.predict()` in scikit-learn. 

In [ ]:
# Your code here

Now, using the formulas of R-squared and adjusted R-squared below, and your Python/numpy knowledge, compute them and contrast them with the R-squared and adjusted R-squared in your statsmodels output using stepwise selection. Which of the two models would you prefer?

$SS_{residual} = \sum (y - \hat{y})^2 $

$SS_{total} = \sum (y - \bar{y})^2 $

$R^2 = 1- \dfrac{SS_{residual}}{SS_{total}}$

$R^2_{adj}= 1-(1-R^2)\dfrac{n-1}{n-p-1}$

In [ ]:
# Your code here

# r_squared is 0.742981  
# adjusted_r_squared is 0.740411

## Level up (Optional)

- Perform variable selection using forward selection, using this resource: https://planspace.org/20150423-forward_selection_with_statsmodels/. Note that this time features are added based on the adjusted R-squared!
- Tweak the code in the `stepwise_selection()` function written above to just perform forward selection based on the p-value 

## Summary
Great! You practiced your feature selection skills by applying stepwise selection and recursive feature elimination to the Boston Housing dataset! 